MinMaxGameHashed.java

```diff
@@ -1,76 +1,90 @@
-		// computer player wants to win as quickly as possible or survive as long 
-		// as it can.
+	@Override
+	protected ValueWithMove getBestMove() {
+		// Getting best move at a current state of the table.
+		// RETURNS: Value of position and best move realizing that value.
 
-		int key = arrayToInt(table);
-		int[] entry = dataBase.get(key);
-		if (entry == null) {
-			boolean lost = (Game.whoWon(table, lastMoveY, lastMoveX, 3-player, lengthToWin) 
-																			== 3-player);
-			if (lost) {
-				int[] value = new int[] {-1, depth};
-				dataBase.put(key, value);
-				return value;
-			}
-			else if (Game.isFull(table)) {
-				int[] value = new int[] {0, depth};
-				dataBase.put(key, value);
-				return value;
-			}
-			else {
-				int[] bestValue = new int[] {-10, depth};
-				int[] value = new int[2];
-				for(int i = 0; i < table.length; i++) {
-					for(int j = 0; j < table[i].length; j++) {
-						if (table[i][j] == 0) {
-							// Trying out move (i,j)
-							table[i][j] = player;
-							value = getValue(table, i, j, 3-player, lengthToWin, depth + 1);
-							table[i][j] = 0;
-							if ((-value[0] == bestValue[0] && bestValue[0] == 1 && 
-															value[1] < bestValue[1]) ||
-								(-value[0] == bestValue[0] && bestValue[0] == -1 && 
-															value[1] > bestValue[1]) ||
-								(-value[0] > bestValue[0])) {
-								
-								bestValue[0] = -value[0];
-								bestValue[1] = value[1];
-							}
-						}
+		
+		// If it was computed before
+		int key = genKey();
+		if (dataBase == null) {
+			dataBase = new HashMap<Integer, int[]>(1000000);
+		}
+		if (dataBase.containsKey(key)) {
+			int[] entry = dataBase.get(key);
+			return new ValueWithMove(entry[0], entry[1], entry[2], entry[3]);
+		}
+
+		// If it needs to be computed
+
+		int tempAgent = activeAgent;
+		int tempY = lastMoveY;
+		int tempX = lastMoveX;
+		
+		activeAgent = 3 - tempAgent;
+		if (winOnLastMove()) {
+			dataBase.put(key, new int[] {-1, movesCompleted, -1, -1});
+			return new ValueWithMove(-1, movesCompleted, -1, -1);
+		}
+		else if (isFull()) {
+			ValueWithMove bestMove = new ValueWithMove(0, movesCompleted, -1, -1);
+			dataBase.put(key, new int[] {0, movesCompleted, -1, -1});
+			return new ValueWithMove(0, movesCompleted, -1, -1);
+		}
+		else {
+			ValueWithMove move;
+			ValueWithMove bestMove = new ValueWithMove(-2,0, -1, -1);
+			for(int i = 0; i < table.length; i++) {
+				for(int j = 0; j < table[i].length; j++) {
+					if (table[i][j] == 0) {
+
+						// Trying out move (i,j)
+						lastMoveY = i;
+						lastMoveX = j;
+						table[i][j] = tempAgent;
+						activeAgent = 3 - tempAgent;
+						movesCompleted += 1;
+
+						// Evaluating the value of the action
+						move = getBestMove();
+						move.flipValue();
+
+						// Undoing tried move
+						table[i][j] = 0;
+						movesCompleted -= 1;
+
+						// Cosmetics on the move
+						move.bestMoveY = i;
+						move.bestMoveX = j;
+
+						// Comparing to current best move
```

Comment:
Not needed as the code is self-explanatory

Reply:
got rid of it